File: converter-python/app/converter.py
"""
LibreOffice subprocess wrapper with async concurrency control.

The public interface is a single coroutine, :func:`convert`, which:

1. Acquires a slot from the shared :class:`asyncio.Semaphore` (bounded by
   ``OFFICE_MAX_CONCURRENT_CONVERSIONS``).
2. Creates an isolated LibreOffice user-profile directory so parallel
   conversions never interfere with each other.
3. Runs ``soffice --headless --convert-to <format>`` under ``coreutils timeout``
   so runaway processes are killed after ``OFFICE_CONVERSION_TIMEOUT``.
4. Releases the slot and cleans up the profile directory unconditionally in a
   ``finally`` block.

The semaphore is lazily initialised inside the running event loop via
:func:`get_semaphore`.  The entrypoint starts uvicorn with ``--workers 1``
so the semaphore is the sole global concurrency cap.  If multiple workers
were used, each would have its own semaphore of size
``max_concurrent_conversions``, giving ``N × cap`` total slots.
"""

import asyncio
import os
import secrets
import shutil
import tempfile

from app.config import ACQUIRE_TIMEOUT, get_settings

# ---------------------------------------------------------------------------
# MIME type map for supported output formats
# ---------------------------------------------------------------------------

CONTENT_TYPES: dict[str, str] = {
    'xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'ods': 'application/vnd.oasis.opendocument.spreadsheet',
    'xlsb': 'application/vnd.ms-excel.sheet.binary.macroEnabled.12',
}

# ---------------------------------------------------------------------------
# Semaphore — one per event loop, lazily created
# ---------------------------------------------------------------------------

_semaphore: asyncio.Semaphore | None = None


def get_semaphore() -> asyncio.Semaphore:
    """Return the process-wide conversion semaphore, creating it on first call.

    The semaphore size is taken from :attr:`~app.config.Settings.max_concurrent_conversions`.
    Lazy initialisation ensures it is created inside the running event loop,
    which is required for ``asyncio.Semaphore`` to work correctly.
    """
    global _semaphore
    if _semaphore is None:
        _semaphore = asyncio.Semaphore(get_settings().max_concurrent_conversions)
    return _semaphore


# ---------------------------------------------------------------------------
# Public coroutine
# ---------------------------------------------------------------------------

async def convert(src: str, out_dir: str, to_format: str) -> str:
    """Convert a spreadsheet file using LibreOffice in headless mode.

    Acquires a concurrency slot, runs ``soffice``, then releases the slot and
    removes the temporary LibreOffice profile directory — even on failure.

    Args:
        src: Absolute path to the source spreadsheet (e.g. ``/tmp/foo/report.xlsb``).
        out_dir: Directory where LibreOffice will write the output file.
        to_format: Target format — ``'xlsx'`` or ``'ods'``.

    Returns:
        Absolute path to the converted output file inside *out_dir*.

    Raises:
        ValueError: *to_format* is not ``'xlsx'`` or ``'ods'``.
        RuntimeError: All conversion slots were busy for :data:`~app.config.ACQUIRE_TIMEOUT`
            seconds (message starts with ``"too many concurrent conversions"``).
        RuntimeError: ``soffice`` exited with a non-zero code.
        RuntimeError: The expected output file was not found after a successful
            ``soffice`` run (should not happen in practice).
    """
    if to_format not in ('xlsx', 'ods'):
        raise ValueError(f"unsupported target format: {to_format!r}")

    sem = get_semaphore()
    try:
        await asyncio.wait_for(sem.acquire(), timeout=ACQUIRE_TIMEOUT)
    except asyncio.TimeoutError:
        raise RuntimeError("too many concurrent conversions")

    settings = get_settings()
    profile = os.path.join(tempfile.gettempdir(), f'lo-profile-{secrets.token_hex(8)}')
    os.makedirs(profile, mode=0o700)

    try:
        output_path = await _run_soffice(src, out_dir, to_format, profile, settings.conversion_timeout_seconds)
    finally:
        shutil.rmtree(profile, ignore_errors=True)
        sem.release()

    return output_path


async def _run_soffice(src: str, out_dir: str, to_format: str, profile: str, timeout_seconds: int) -> str:
    """Spawn ``soffice`` as an async subprocess and wait for it to finish.

    Uses ``coreutils timeout`` with a 5-second SIGKILL grace period so that
    a runaway LibreOffice process is always cleaned up.

    Args:
        src: Path to the source file.
        out_dir: Directory for LibreOffice output.
        to_format: Target format string (``'xlsx'`` or ``'ods'``).
        profile: Path to the isolated LibreOffice user-profile directory.
        timeout_seconds: Hard deadline in seconds; soffice is SIGTERM'd at this
            point and SIGKILL'd 5 seconds later.

    Returns:
        Absolute path to the generated output file.

    Raises:
        RuntimeError: ``soffice`` exited non-zero or the output file is missing.
    """
    cmd = [
        'timeout', '--foreground', '--kill-after=5s', f'{timeout_seconds}s',
        'soffice', '--headless', '--nologo', '--nofirststartwizard',
        f'-env:UserInstallation=file://{profile}',
        '--convert-to', to_format,
        '--outdir', out_dir,
        src,
    ]

    proc = await asyncio.create_subprocess_exec(
        *cmd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate()

    if proc.returncode != 0:
        output = (stdout + stderr).decode(errors='replace').strip()
        raise RuntimeError(f"conversion failed (exit {proc.returncode}): {output}")

    stem = os.path.splitext(os.path.basename(src))[0]
    dst = os.path.join(out_dir, f'{stem}.{to_format}')
    if not os.path.isfile(dst):
        raise RuntimeError(f"expected output file was not generated: {dst}")

    return dst

File: converter-php/api/src/Converter.php
<?php
declare(strict_types=1);

namespace App;

/**
 * Converter
 *
 * LibreOffice headless conversion + concurrency control.
 * (Logic originally ported from the Go version; now used by the Slim app.)
 */
class Converter
{
    private static ?int $maxConcurrent = null;
    private static ?int $timeoutSeconds = null;

    private static function getMaxConcurrent(): int
    {
        if (self::$maxConcurrent === null) {
            self::$maxConcurrent = max(1, (int)(getenv('OFFICE_MAX_CONCURRENT_CONVERSIONS') ?: '2'));
        }
        return self::$maxConcurrent;
    }

    private static function getTimeoutSeconds(): int
    {
        if (self::$timeoutSeconds === null) {
            $d = strtolower(trim(getenv('OFFICE_CONVERSION_TIMEOUT') ?: '60s'));
            if (preg_match('/^(\d+)(s|m|h)?$/', $d, $m)) {
                $val = (int)$m[1];
                $unit = $m[2] ?? 's';
                if ($unit === 'm') $val *= 60;
                elseif ($unit === 'h') $val *= 3600;
                self::$timeoutSeconds = $val;
            } else {
                self::$timeoutSeconds = 60;
            }
        }
        return self::$timeoutSeconds;
    }

    /**
     * Acquire one of N conversion slots using non-blocking flock.
     * Returns open file resource or null on timeout.
     */
    private static function acquireSlot(): mixed
    {
        $max = self::getMaxConcurrent();
        $start = microtime(true);
        $waitMax = 30.0;

        while (true) {
            for ($i = 0; $i < $max; $i++) {
                $path = sys_get_temp_dir() . "/.lo-slot-{$i}.lock";
                $fp = @fopen($path, 'c+');
                if ($fp && flock($fp, LOCK_EX | LOCK_NB)) {
                    return $fp;
                }
                if ($fp) {
                    @fclose($fp);
                }
            }
            if (microtime(true) - $start > $waitMax) {
                return null;
            }
            usleep(150 * 1000);
        }
    }

    private static function releaseSlot($fp): void
    {
        if (is_resource($fp)) {
            @flock($fp, LOCK_UN);
            @fclose($fp);
        }
    }

    public static function rrmdir(string $dir): void
    {
        if (!is_dir($dir)) {
            return;
        }
        $it = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($dir, \FilesystemIterator::SKIP_DOTS),
            \RecursiveIteratorIterator::CHILD_FIRST
        );
        foreach ($it as $item) {
            if ($item->isDir()) {
                @rmdir($item->getPathname());
            } else {
                @unlink($item->getPathname());
            }
        }
        @rmdir($dir);
    }

    /**
     * Convert the source file to target format (xlsx or ods).
     * Returns absolute path to the generated file.
     * Throws \RuntimeException on failure.
     */
    public static function convertTo(string $src, string $outDir, string $toFormat): string
    {
        if ($toFormat !== 'xlsx' && $toFormat !== 'ods') {
            throw new \RuntimeException("unsupported target format: $toFormat (use xlsx or ods)");
        }

        $lock = self::acquireSlot();
        if ($lock === null) {
            throw new \RuntimeException("too many concurrent conversions");
        }

        $profile = sys_get_temp_dir() . '/lo-profile-' . bin2hex(random_bytes(8));
        if (!@mkdir($profile, 0700, true)) {
            self::releaseSlot($lock);
            throw new \RuntimeException("could not create temporary profile");
        }

        $timeout = self::getTimeoutSeconds();

        $cmd = [
            'timeout', '--foreground', '--signal=TERM', '--kill-after=5s', $timeout . 's',
            'soffice', '--headless',
            '--nologo',
            '--nofirststartwizard',
            '-env:UserInstallation=file://' . $profile,
            '--convert-to', $toFormat,
            '--outdir', $outDir,
            $src,
        ];

        $descriptors = [
            0 => ['pipe', 'r'],
            1 => ['pipe', 'w'],
            2 => ['pipe', 'w'],
        ];
        $pipes = [];

        $proc = proc_open($cmd, $descriptors, $pipes);
        if (!is_resource($proc)) {
            self::rrmdir($profile);
            self::releaseSlot($lock);
            throw new \RuntimeException("failed to start soffice process");
        }

        fclose($pipes[0]);
        $stdout = stream_get_contents($pipes[1]);
        $stderr = stream_get_contents($pipes[2]);
        fclose($pipes[1]);
        fclose($pipes[2]);

        $exitCode = proc_close($proc);
        $combined = trim($stdout . "\n" . $stderr);

        self::rrmdir($profile);
        self::releaseSlot($lock);

        if ($exitCode !== 0) {
            throw new \RuntimeException("conversion failed (exit $exitCode): $combined");
        }

        $base = basename($src);
        $ext = pathinfo($base, PATHINFO_EXTENSION);
        $stem = $ext !== '' ? substr($base, 0, -strlen('.' . $ext)) : $base;
        $dst = rtrim($outDir, '/') . '/' . $stem . '.' . $toFormat;

        if (!is_file($dst)) {
            throw new \RuntimeException("expected $toFormat was not generated: $dst");
        }

        return $dst;
    }

    public static function convertXlsb(string $src, string $outDir): string
    {
        return self::convertTo($src, $outDir, 'xlsx');
    }
}
